### PR TITLE
[trello.com/c/bSmPAIM2] Feat: Show fixed date on scroll

### PR DESCRIPTION
**Pull request by IanaaDvlp**

### CODE REVIEW

File: Adamant/Modules/Chat/ViewModel/ChatViewModel.swift [hunk 1]
```diff
@@ -1023,6 +1037,13 @@ extension ChatViewModel: NSFetchedResultsControllerDelegate {
 }
 
 private extension ChatViewModel {
+    func startHideDateTimer() {
+        dateTimer?.invalidate()
+        dateTimer = Timer.scheduledTimer(withTimeInterval: 2.0, repeats: false) { [weak self] _ in
```
Comment (just-software-dev) [inline Adamant/Modules/Chat/ViewModel/ChatViewModel.swift:1042]: Use Combine

File: Adamant/Modules/Chat/View/ChatViewController.swift [hunk 2]
```diff
@@ -189,11 +194,29 @@ final class ChatViewController: MessagesViewController {
         super.collectionView(collectionView, willDisplay: cell, forItemAt: indexPath)
     }
     
+    override func scrollViewDidEndDragging(_: UIScrollView, willDecelerate _: Bool) {
```
Comment (just-software-dev) [inline Adamant/Modules/Chat/View/ChatViewController.swift:197]: Are you sure it's correct? At this point scroll can still have momentum and can actually stop a bit later

File: Adamant/Modules/Chat/View/ChatViewController.swift [hunk 3]
```diff
@@ -189,11 +194,29 @@ final class ChatViewController: MessagesViewController {
...
+        }) as [IndexPath]? else { return }
+        for indexPath in visibleIndexPaths {
+            if let cell = messagesCollectionView.cellForItem(at: indexPath) {
+                let cellRect = messagesCollectionView.convert(cell.frame, to: self.view)
+                if cellRect.minY <= targetY && cellRect.maxY >= targetY {
+                    viewModel.checkTopMessage(indexPath: indexPath)
+                    break
+                }
+            }
+        }
```
Comment (just-software-dev) [inline Adamant/Modules/Chat/View/ChatViewController.swift:218]: Move it to a separate method like `updateScrollDownButtonVisibility()`

File: Adamant/Modules/Chat/ViewModel/ChatViewModel.swift [hunk 4]
```diff
@@ -1014,6 +1017,17 @@ extension ChatViewModel {
...
+    
+    func checkTopMessage(indexPath: IndexPath) {
+        guard let message = messages[safe: indexPath.section],
+              let date = message.dateHeader?.string.string
+        else { return }
+        dateHeader = date
+        dateHeaderHidden = false
+    }
+    func didEndScroll() {
+        startHideDateTimer()
```
Comment (just-software-dev) [inline Adamant/Modules/Chat/ViewModel/ChatViewModel.swift:1029]: Missed spacing between these two methods

File: Adamant/Modules/Chat/ViewModel/ChatViewModel.swift [hunk 5]
```diff
@@ -1014,6 +1017,17 @@ extension ChatViewModel {
...
     }
+    
+    func checkTopMessage(indexPath: IndexPath) {
+        guard let message = messages[safe: indexPath.section],
+              let date = message.dateHeader?.string.string
+        else { return }
+        dateHeader = date
+        dateHeaderHidden = false
+    }
+    func didEndScroll() {
```
Comment (just-software-dev) [inline Adamant/Modules/Chat/ViewModel/ChatViewModel.swift:1028]: Missed spacing between these two methods

File: Adamant/Modules/Chat/View/ChatViewController.swift [hunk 6]
```diff
@@ -492,6 +525,16 @@ private extension ChatViewController {
...
+        
+        view.addSubview(dateHeaderLabel)
+        dateHeaderLabel.backgroundColor = .adamant.chatSenderBackground
+        dateHeaderLabel.textInsets = .init(top: 4, left: 7, bottom: 4, right: 7)
+        dateHeaderLabel.layer.cornerRadius = 10
+        dateHeaderLabel.clipsToBounds = true
+        dateHeaderLabel.snp.makeConstraints { make in
+            make.top.equalTo(view.safeAreaLayoutGuide).offset(10)
+            make.centerX.equalToSuperview()
+        }
```
Comment (just-software-dev) [inline Adamant/Modules/Chat/View/ChatViewController.swift:537]: Avoid defining views in huge view controller. There is too many code in this file. I suggest to move it to the other file

File: Adamant/Modules/Chat/View/ChatViewController.swift [hunk 7]
```diff
@@ -189,11 +194,29 @@ final class ChatViewController: MessagesViewController {
...
+    override func scrollViewDidEndDragging(_: UIScrollView, willDecelerate _: Bool) {
+        viewModel.didEndScroll()
+    }
+    
     override func scrollViewDidScroll(_ scrollView: UIScrollView) {
         super.scrollViewDidScroll(scrollView)
         updateIsScrollPositionNearlyTheBottom()
         updateScrollDownButtonVisibility()
         
+        let targetY: CGFloat = 20 + view.safeAreaInsets.top
```
Comment (StanislavDevIOS) [inline Adamant/Modules/Chat/View/ChatViewController.swift:206]: move `20` to constants. Connect it with `dateHeaderLabel`

File: Adamant/Modules/Chat/View/ChatViewController.swift (hunk 3 above)
Comment (StanislavDevIOS) [reply Adamant/Modules/Chat/View/ChatViewController.swift:218]: use `guard` to improve code readability

File: Adamant/Modules/Chat/ViewModel/ChatViewModel.swift (hunk 1 above)
Comment (StanislavDevIOS) [reply Adamant/Modules/Chat/ViewModel/ChatViewModel.swift:1042]: move `2.0` to constants

**Pull request merged**